test/unit/org/apache/cassandra/index/sai/plan/EstimatedRowCountTest.java

```diff
@@ -0,0 +1,126 @@
+import org.apache.cassandra.index.sai.SAIUtil;
+import org.apache.cassandra.index.sai.disk.format.Version;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.fail;
+
+public class EstimatedRowCountTest extends SAITester
+{
+    private int queryOptLevel;
+
+    @Before
+    public void setup()
+    {
+        queryOptLevel = QueryController.QUERY_OPT_LEVEL;
+        QueryController.QUERY_OPT_LEVEL = 0;
+    }
+
+    @After
+    public void teardown()
+    {
+        QueryController.QUERY_OPT_LEVEL = queryOptLevel;
+    }
+
+    private Object getFilterValue(String type, int value)
+    {
+        switch (type)
+        {
+            case "int":
+                return value;
+            case "decimal":
+                return BigDecimal.valueOf(value);
+            case "varint":
+                return BigInteger.valueOf(value);
+        }
+        fail("Must be known type");
+        return null;
+    }
+
+    private ColumnFamilyStore prepareTable(String type)
+    {
+        createTable("CREATE TABLE %s (pk text PRIMARY KEY, age " + type + ')');
+        createIndex("CREATE CUSTOM INDEX ON %s(age) USING 'StorageAttachedIndex'");
+        for (int i = 0; i < 100; i++)
+        {
+            execute("INSERT INTO %s (pk, age) VALUES (?," + i + ')', "key" + i);
+        }
+        return getCurrentColumnFamilyStore();
+    }
+
+    private void doInequalityTest(Version version, String type, double expectedRows)
+    {
+        Version latest = Version.latest();
+        SAIUtil.setLatestVersion(version);
+
+        ColumnFamilyStore cfs = prepareTable(type);
+        double totalRows = 97.0;
+        Object filter = getFilterValue(type, 25);
+
+        ReadCommand rc = Util.cmd(cfs)
+                             .columns("age")
+                             .filterOn("age", Operator.NEQ, filter)
+                             .build();
+        QueryController controller = new QueryController(cfs,
+                                                         rc,
+                                                         version.onDiskFormat().indexFeatureSet(),
+                                                         new QueryContext(),
+                                                         null);
+
+        Plan plan = controller.buildPlan();
+        assert plan instanceof Plan.RowsIteration;
+        Plan.RowsIteration root = (Plan.RowsIteration) plan;
+        Plan.KeysIteration planNode = root.firstNodeOfType(Plan.KeysIteration.class);
+        assertNotNull(planNode);
+
+        assertEquals(expectedRows, root.expectedRows(), 0.1);
+        assertEquals(expectedRows, planNode.expectedKeys(), 0.1);
+        assertEquals(expectedRows / totalRows, planNode.selectivity(), 0.001);
+
+        SAIUtil.setLatestVersion(latest);
+    }
+
+    @Test
+    public void testInequality()
+    {
+        doInequalityTest(Version.DB, "int", 97.0);
```

Comment:
This param 97.0 seems to be not doing anything. If I set it to a different value, then the test fails.
Are you sure you need this to be a param?

Reply:
When I implemented Anti-join in #1439 it was different estimation for anti-join plan. I guess it will be different when I introduce union-based implementation.
I will add a comment, since it just happens that it's equivalent now.

Reply:
I added a test, which has different estimations depending if histogram is used or not.